### PR TITLE
Tests for and fixes to all existing Date.java functionlity

### PR DESCRIPTION
**Pull request by R055A**
# Requirements

    Python v3.85
    Java v8
    Junit v5
    JaCoCo v0.8.6

# Instructions

    - Open the voc/ directory as a project in IntelliJ
    - In File -> Project Structure -> Project -> Project SDK select Java 1.8
    - It may be required to disable android. If so, also in Project Settings, select Modules -> Sources, right click python/android and select Excluded
    - Navigate to voc/python/test/python/stdlib/datetime/
    - Junit5 may require being installed. If so, right click on any junit import, select Show Context Actions and select junit5
    - In Edit Configurations... -> Add New Configuration (the + symbol), select Junit
    - JaCoCo can be selected in Edit Configurations.. -> Code Coverage -> Choose Coverage Runner
    - Either right-click in the test file or on the directory to run all 56 tests or click the green run button on the left of each test to run them individually

# Tested functionality (following any and all fixes and refactoring)

#### Instance attributes
   
    + year: final Int
    + month: final Int
    + day: final Int

#### Class attributes
    
    + min: Date() <<static>>
    + max: Date() <<static>>

#### Instance methods

    + Date(Object[], Map<String, Object>): Date()
    + __repr__(): Str
    + ctime(): Object
    + weekday(): Int

#### Class methods

    + today(): Date() <<static>>

#### Private methods (not tested directly)

    - validateThreeInputs(): void
    - validateTwoInputs(): void
    - validateOneInput(): void
    - validateArgTypes(): void
    - validateKwargTypes(): void
    - validateAttributeRanges(): void
    - __max__(): Date() <<static>>
    - __min__(): Date() <<static>>

# Removed functionality

    + constant_4(): Object <<static>>
    + __year__(): Str
    + __month__(): Str
    + __day__(): Str

# Notes

Testing is focused on all of the different possible outputs when using the Python datetime.date module for all Python types and value ranges for each different variation between args and kwargs input for each different number of inputs in total.

It was found not possible to implement class attributes and methods with the same name as instance attributes and methods. In each case it was determined which between instance and class had the most significance in functionality and the least of the two was either removed or not implemented. For example, the instance attributes year, month and day are extremely significant, whereas the class attributes only return a string, such as:  <attribute 'year' of 'datetime.date' objects>. Thus, the public methods \__year\__(), \__month\__() and \__day\__(), each returning a string, where removed, although they have underscores on either side of the name, which in itself is reason to remove them because Python does not implement such named methods. In the case of date.today(), this returns a new instance for the current date and no other instance should ever be required so only the class method for this is used in the Date() class.

There are four disabled tests. These are for testing functionality that can not be implemented in the Date() java class. This is because the input arguments are split into args and kwargs in the java implementation, whereas in Python the input args are not split. Thus, in Python the order of input between keyword and name arguments can be known, but in the java Date() class the order of input between args and kwargs can not be known.

### CODE REVIEW

**Comment by DavidOstlund** [comment]
## Review
The code looks good and as you mentioned the code coverage is really high at 100% class coverage, 94% method coverage and 97% line coverage. We have some comments about how to make the code more readable. 

### Date.java
There are some long one liners on line 384 and line 344+345.
We believe that breaking them up over several assignments would increase readability.

Some comments explaining the if statement starting on line 368 and ending on line 374 would help with understanding what's going on.
It was a bit hard to see what the non variable numbers represented by just reading the code.

Out of curiosity we were also wondering in what order you did your tests.
Did you first create them and then do coverage, or the other way around?
We have not looked at every test in detail but we were wondering if might be possible to get a similar amount of coverage with less tests.

Anyhow, really great job.
Well done!

**Comment by R055A** [comment]
> 
> ## Review
> 
> The code looks good and as you mentioned the code coverage is really high at 100% class coverage, 94% method coverage and 97% line coverage. We have some comments about how to make the code more readable.
> ### Date.java
> 
> There are some long one liners on line 384 and line 344+345.
> We believe that breaking them up over several assignments would increase readability.
> 
> Some comments explaining the if statement starting on line 368 and ending on line 374 would help with understanding what's going on.
> It was a bit hard to see what the non variable numbers represented by just reading the code.
> 
> Out of curiosity we were also wondering in what order you did your tests.
> Did you first create them and then do coverage, or the other way around?
> We have not looked at every test in detail but we were wondering if might be possible to get a similar amount of coverage with less tests.
> 
> Anyhow, really great job.
> Well done!

Thank you!

The tests are created first. Then the methods are fixed so that they pass the testing for them. The coverage is a measure of how well the methods are tested, if all the different possible implemented outcomes are considered in the testing, and is only relevant when they are tested. Thus, the tests must come first. The methods being tested require having all tested functionalities implemented to pass all tests. If all tests are passing then the methods have all this functionality and, unless the methods have additional irrelevant implementations, the coverage should always be high. If any tests are removed, then not all functionality implemented to pass these tests will be covered by the testing and the coverage report will show this with lower percentages. 

When you run code coverage it is possible to also check for branch coverage. This can be done in IntelliJ by selecting tracing in the code coverage options. The branch coverage when last run was above 80%, but you should check for yourself to be sure.

As for lines 344 and 345, the attributes for each of year, month and day from the LocalDate.now() instance are being converted to org.python.types.Int type in an Object[] array so they can be passed as args in a new Date() instance being returned in the next line. This will be broken up to help with readability as requested very soon.

As for line 384, this is a solution to the weekday() method bug. First, this.weekdays is a HashMap that maps each weekday String return by the java LocalDate().getDayOfWeek() method to its corresponding weekday number returned by the Python datetime.date().weekday() method. The HashMap is initialized as a global, and can be found declared at line #50 and initialized and set from lines #126 to #133. To get the weekday String, it is required to use the LocalDate.parse() method. To get a string to enter as an argument to this method, either the datetime.date().str() method is required or, otherwise, the inbuilt __repr__() method, which is what is being given as the argument to LocalDate.parse(). The __repr__() method can be found at lines #297 to #311. This can also be broken up to help with readability as requested, which will be done very soon.

I hope that that answered your questions adequately. 

As for lines 368 to 374, this will require being answered by @nguyenluc99, the collaborator responsible for the bug fix.

**Review by nguyenluc99 — CHANGES_REQUESTED** [review]
Just a small change in checking a leap year.
Leap year is a year which is divided by 4, not divided by 100; or divided by 400.

File: python/common/org/python/stdlib/datetime/Date.java [hunk 1]
```diff
@@ -1,267 +1,387 @@
...
+
+        // Check month attribute value is within the appropriate range
+        if (this.minDate > this.month.value || this.month.value > this.maxMonth) {
+            throw new org.python.exceptions.ValueError(DateTimeEnum.MON_VAL_ERR.toString());
+        }
+        int maxDayRange;
+
+        // Get the maximum day range for the given month and year
+        if (this.month.value == 2) {
+            if (this.year.value % 4 == 0) {
```
[the remark this answers is not in the document]
Comment (R055A) [reply python/common/org/python/stdlib/datetime/Date.java:275]: These are not the lines of code being referred to by @DavidOstlund

File: python/common/org/python/stdlib/datetime/Date.java [hunk 2]
```diff
@@ -1,267 +1,387 @@
...
+        long yearNum = ((org.python.types.Int) this.year).value;
+        String yearStr = new String(Long.toString(yearNum));
+        if (yearNum < 10) {
+            yearStr = "000" + yearStr;
+        } else if (yearNum < 100) {
+            yearStr = "00" + yearStr;
+        } else if (yearNum < 1000) {
+            yearStr = "0" + yearStr;
+        }
+        return new org.python.types.Str(weekdayStr + " " + monthStr + " " + dayStr + " 00:00:00 " + yearStr);
```
[the remark this answers is not in the document]
Comment (R055A) [reply python/common/org/python/stdlib/datetime/Date.java:375]: @DavidOstlund wants to understand what the code does between lines 368 and 374. He is requesting some line comments.

**Comment by DavidOstlund** [comment]
> > ## Review
> > The code looks good and as you mentioned the code coverage is really high at 100% class coverage, 94% method coverage and 97% line coverage. We have some comments about how to make the code more readable.
> > ### Date.java
> > There are some long one liners on line 384 and line 344+345.
> > We believe that breaking them up over several assignments would increase readability.
> > Some comments explaining the if statement starting on line 368 and ending on line 374 would help with understanding what's going on.
> > It was a bit hard to see what the non variable numbers represented by just reading the code.
> > Out of curiosity we were also wondering in what order you did your tests.
> > Did you first create them and then do coverage, or the other way around?
> > We have not looked at every test in detail but we were wondering if might be possible to get a similar amount of coverage with less tests.
> > Anyhow, really great job.
> > Well done!
> 
> Thank you!
> 
> The tests are created first. Then the methods are fixed so that they pass the testing for them. The coverage is a measure of how well the methods are tested, if all the different possible implemented outcomes are considered in the testing, and is only relevant when they are tested. Thus, the tests must come first. The methods being tested require having all tested functionalities implemented to pass all tests. If all tests are passing then the methods have all this functionality and, unless the methods have additional irrelevant implementations, the coverage should always be high. If any tests are removed, then not all functionality implemented to pass these tests will be covered by the testing and the coverage report will show this with lower percentages.
> 
> When you run code coverage it is possible to also check for branch coverage. This can be done in IntelliJ by selecting tracing in the code coverage options. The branch coverage when last run was above 80%, but you should check for yourself to be sure.
> 
> As for lines 344 and 345, the attributes for each of year, month and day from the LocalDate.now() instance are being converted to org.python.types.Int type in an Object[] array so they can be passed as args in a new Date() instance being returned in the next line. This will be broken up to help with readability as requested very soon.
> 
> As for line 384, this is a solution to the weekday() method bug. First, this.weekdays is a HashMap that maps each weekday String returned by the java LocalDate().getDayOfWeek() method to its corresponding weekday number returned by the Python datetime.date().weekday() method. The HashMap is initialized as a global, and can be found declared at line 50 and initialized and set from lines 126 to 133. To get the weekday String, it is required to use the LocalDate.parse() method. To get a string to enter as an argument to this method, either the datetime.date().__str__() method is required or, otherwise, the inbuilt __repr__() method, which is what is being given as the argument to LocalDate.parse(). The __repr__() method can be found at lines 297 to 311. Line 384 can also be broken up to help with readability as requested, which will be done very soon.
> 
> I hope that that answered your questions adequately.
> 
> As for lines 368 to 374, this will require being answered by @nguyenluc99, the contributor responsible for the bug fix.

**Comment by DavidOstlund** [comment]
> > ## Review
> > The code looks good and as you mentioned the code coverage is really high at 100% class coverage, 94% method coverage and 97% line coverage. We have some comments about how to make the code more readable.
> > ### Date.java
> > There are some long one liners on line 384 and line 344+345.
> > We believe that breaking them up over several assignments would increase readability.
> > Some comments explaining the if statement starting on line 368 and ending on line 374 would help with understanding what's going on.
> > It was a bit hard to see what the non variable numbers represented by just reading the code.
> > Out of curiosity we were also wondering in what order you did your tests.
> > Did you first create them and then do coverage, or the other way around?
> > We have not looked at every test in detail but we were wondering if might be possible to get a similar amount of coverage with less tests.
> > Anyhow, really great job.
> > Well done!
> 
> Thank you!
> 
> The tests are created first. Then the methods are fixed so that they pass the testing for them. The coverage is a measure of how well the methods are tested, if all the different possible implemented outcomes are considered in the testing, and is only relevant when they are tested. Thus, the tests must come first. The methods being tested require having all tested functionalities implemented to pass all tests. If all tests are passing then the methods have all this functionality and, unless the methods have additional irrelevant implementations, the coverage should always be high. If any tests are removed, then not all functionality implemented to pass these tests will be covered by the testing and the coverage report will show this with lower percentages.
> 
> When you run code coverage it is possible to also check for branch coverage. This can be done in IntelliJ by selecting tracing in the code coverage options. The branch coverage when last run was above 80%, but you should check for yourself to be sure.
> 
> As for lines 344 and 345, the attributes for each of year, month and day from the LocalDate.now() instance are being converted to org.python.types.Int type in an Object[] array so they can be passed as args in a new Date() instance being returned in the next line. This will be broken up to help with readability as requested very soon.
> 
> As for line 384, this is a solution to the weekday() method bug. First, this.weekdays is a HashMap that maps each weekday String returned by the java LocalDate().getDayOfWeek() method to its corresponding weekday number returned by the Python datetime.date().weekday() method. The HashMap is initialized as a global, and can be found declared at line 50 and initialized and set from lines 126 to 133. To get the weekday String, it is required to use the LocalDate.parse() method. To get a string to enter as an argument to this method, either the datetime.date().__str__() method is required or, otherwise, the inbuilt __repr__() method, which is what is being given as the argument to LocalDate.parse(). The __repr__() method can be found at lines 297 to 311. Line 384 can also be broken up to help with readability as requested, which will be done very soon.
> 
> I hope that that answered your questions adequately.
> 
> As for lines 368 to 374, this will require being answered by @nguyenluc99, the contributor responsible for the bug fix.

Thank you for the answers.
Now I understand what's going on :)

File: python/common/org/python/stdlib/datetime/Date.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nguyenluc99) [reply python/common/org/python/stdlib/datetime/Date.java:379]: If the year is less than 10, for example yearNum=5, the previous code will print like ...00:00:00 5, but python3 print it like ...00:00:00 0005, so we need to add extra "000" before printing the year. If it is not less than 10, it should be check for being less than 100, to add only "00" in front of it. Similar to the case of being less than 1000.

**Review by nguyenluc99 — APPROVED** [review]
so nice

**Review by DavidOstlund — APPROVED** [review]

**Review by ponterik — APPROVED** [review]

**Pull request merged**